Initial draft of contributing guide (#1787)
madolson/placeholderkv: DEVELOPMENT_GUIDE.md

# Valkey development guidelines

> This document provides an general overview for writing and designing code for Valkey.

Author: ```suggestion
This document provides a general overview for writing and designing code for Valkey.
```

During our long development history, we've made a lot of inconsistent decisions, but we strive to get incrementally better.

## General style guidelines
Most of the style guidelines are enforced by clang format, but some additional comments are included here.

1. C style comments `/* comment */` can be used for both single and multi-line comments. C++ comments `//` can only be used for single line comments.

> 1. Generally keep line lengths below 90 characters, however there is no explicit line length enforcement.

Reviewer: I believe we landed on 120'ish previously but I am personally more in favor of 90 .

Author: The original decision was prefer less than 90, but have a strong reason to have less than 120.

> 1. Use static functions when a function is only intended to be accessed from the same file.

Reviewer: should we also outlaw the "_" prefix explicitly now? We can fix the existing ones gradually. 

@zuiderkwast

Author: Sure

## Naming conventions
Valkey has a long history of inconsistent naming conventions. Generally follow the style of the surrounding code, but you can also always use the following conventions for variable and structure names:

> - Variable names: `snake_case` or all lower case (e.g. `valkeyobject` or `valkey_object`)

Reviewer: can we converge on `snake_case` exclusively, for the new code?

Author: I generally don't like, "For new code use X", because than it's confusing if someone is trying to conform to the existing standard. Either we should enforce it everywhere I think we should be clear about A or B.

Reviewer 2: I like the "follow surrounding code" guideline.

The all-lower is common in standard C function names like strlen, memcpy, etc. I think it's OK to use this style for short names like keylen, objptr if surrounding code uses this style, but it becomes hard to read in longer names like valkeyobject.

```suggestion
- Variable names: `snake_case` or all lower case for short names  (e.g. `valkey_object` or `keylen`)
```

- Function names: `camelCase` or `namespace_camelCase` (e.g. `createObjectList` or `networking_createObjectList`).
- Macros: `UPPER_CASE` (e.g. `DICT_CREATE`)
- Structures: `camelCase` (e.g. `user`)

## Licensing information
When creating new source code files, use the following snippet to indicate the license:
```
/*
 * Copyright (c) Valkey Contributors
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 */
```

If you are making material changes to a file that has a different license at the top, also add the above license snippet.
There isn't a well defined test for what is considered a material change, but a good rule of thumb is if it's more than 100 lines of code.

> ## Test coverage

Reviewer 3: is there a documentation that mentions how can a developer run some tests? single tests? loop through tests etc

Reviewer 3: If not, should we mention this here?

Reviewer 4: It would be in the `tests/README` IIRC.
And I think the options for `runtest` like loop, should be added in `tests/README`

Reviewer 3: yeah, makes sense! I can probably spend some time and improve it separately.

Valkey uses two types of tests: unit and integration tests.
All contributions should include a test of some form. 

Unit tests are present in the `src/unit` directory, and are intended to test individual structures or files.

> For example, most changes to datastructures should include corresponding unit tests.

Reviewer: ```suggestion
For example, most changes to data structures should include corresponding unit tests.
```

Integration tests are located in the `tests/` directory, and are intended to test end-to-end functionality.
Adding new commands should come with corresponding integration tests.
When writing cluster mode tests, do not use the legacy `tests/cluster` framework, which has been deprecated.

> 

Reviewer: can we add a documentation section as well? new commands, configs, non-trivial features should come with documentation? blogs are also encouraged for major features, such as the 1M QPS, the hash table rewrite

Author: Yeah, I'll add a blurb here.

# Valkey development guidelines
This document provides a general overview for writing and designing code for Valkey.
During our long development history, we've made a lot of inconsistent decisions, but we strive to get incrementally better.

> ## General style guidelines

Reviewer 5: lets place a comment about prefer use of `bool` over `int` in new added code.

Reviewer 5: and?

Author: lol, I was still thinking about it. I don't know if I care if you use bool or not.

Author: Ok, wrote some guidance.

Most of the style guidelines are enforced by clang format, but some additional comments are included here.

> 1. C style comments `/* comment */` can be used for both single and multi-line comments. C++ comments `//` can only be used for single line comments.

Reviewer 5: We also start multilines with a * (and clang formatter will kill you in case you leave a space after the closing '*/' IIRC) also I recall we once kept that a multiline comment is always ending with the '*/' on the same line and not on a new line like:
```
/* blh blh blh blh
 * blh blh blh blh */
```

and not:

```
/* blh blh blh blh
 * blh blh blh blh 
 */
```

Reviewer 2: Yeah, and some information about when to use comments and when not. For example:

Use commends to explain why something is done the way it is, when it's not clear otherwise.

Don't use comments to describe what the code is doing. That should be clear from the code itself.

1. Generally keep line lengths below 90 characters when reasonable, however there is no explicit line length enforcement.
   Use your best judgement for readability.

> 1. Use static functions when a function is only intended to be accessed from the same file.

Reviewer 6: ```suggestion
1. C style comments `/* comment */` can be used for both single and multi-line comments. C++ comments `//` can only be used for single line comments.
2. Generally keep line lengths below 90 characters when reasonable, however there is no explicit line length enforcement.
   Use your best judgement for readability.
3. Use static functions when a function is only intended to be accessed from the same file.
```

Reviewer 5: >Use static functions when a function is only intended to be accessed from the same file.

when the function is not declared in any header file, should it not be made static? this is somewhat what we require today...

Author: Maybe a preference of mine, but I normally use numbered lists with all 1s (since markdown renders them correctly). That way a new item can be added without having to modify subsequent lines if it wasn't added to the end.

   For historical reasons, some private functions are prefixed by `_`, and they are kept as is to make it easier to backport changes.

## Naming conventions
Valkey has a long history of inconsistent naming conventions.
Generally follow the style of the surrounding code, but you can also always use the following conventions for variable and structure names:

- Variable names: `snake_case` or all lower case for short names  (e.g. `valkey_object` or `keylen`)
- Function names: `camelCase` or `namespace_camelCase` (e.g. `createObjectList` or `networking_createObjectList`).
- Macros: `UPPER_CASE` (e.g. `DICT_CREATE`)

> - Structures: `camelCase` (e.g. `user`)

Reviewer 2: We only use the namespace form in certain areas where it's used concistenly. The `networking` namespace is not one of these, so I don't think it's a good example.

I don't think we should use "valkey" as an example as we've been trying to avoid trademarks in code. We've used "server" instead for many server things.

Let's use examples of existing functions/variables/macros instead of made up ones.

```suggestion
- Variable names: `snake_case` or all lower case for short names  (e.g. `cached_reply` or `keylen`).
- Function names: `camelCase` or `namespace_camelCase` (e.g. `createStringObject` or `IOJobQueue_isFull`).
- Macros: `UPPER_CASE` (e.g. `MAKE_CMD`).
- Structures: `camelCase` (e.g. `user`).
```

## Documentation
Valkey keeps most of the user documentation in the [valkey-doc](https://github.com/valkey-io/valkey-doc) repository in a few areas:
1. Major functionality is documented in the [topics](https://github.com/valkey-io/valkey-doc/tree/main/topics) section.
2. Specific command behavior is documented in the [commands](https://github.com/valkey-io/valkey-doc/tree/main/commands) section.
   Command history is also documented in the [command json file](https://github.com/valkey-io/valkey/tree/unstable/src/commands). 
3. Server info fields are documented in the [info](https://github.com/valkey-io/valkey-doc/blob/main/commands/info.md) command.

When a PR is opened that requires documentation to be updated, the `needs-doc-pr` should be added until the corresponding documentation PR is open.

## Best practices
1. Avoid adding configuration when a feature can be fully controlled by heuristics. 
We want Valkey to work correctly out of the box without much tuning.
Configurations can be added to provide additional tuning of features. 
When the workload characteristics can't be inferred or imply a tradeoff (CPU vs memory), then provide a configuration.

> 2. Try to limit the number of lines changed in a PR when possible.

Reviewer 2: This is the most important guideline of all. 😄 We could place near the top of the page rather than at the bottom.

Author: I'll move best practices and other stuff up to the top!

## General style guidelines
Most of the style guidelines are enforced by clang format, but some additional comments are included here.

1. C style comments `/* comment */` can be used for both single and multi-line comments.
   C++ comments `//` can only be used for single line comments.
   Multi line comments should have the leading `*` align and the final `*/` should be on the same line as the last line of text.
   e.g.
```c
/* Blah Blah
 * Blah Blah. */
   ```
2. Comments should generally be used to describe behavior that is not obvious from reading the code itself.
   This includes complex behavior, why code was written the way it was and describing non-obvious behavior.
   Additionally, functions should be documented to explain all of the function's behavior without having to read the code.
1. Generally keep line lengths below 90 characters when reasonable, however there is no explicit line length enforcement.
   Use your best judgement for readability.
1. Use static functions when a function is only intended to be accessed from the same file.
   For historical reasons, some private functions are prefixed by `_`, and they are kept as is to make it easier to backport changes.

> 

Author: ```suggestion
   For historical reasons, some private functions are prefixed by `_`, and they are kept as is to make it easier to backport changes.
1. Use the boolean type for true/false values.
   For historical reasons, some functions used the integer type, and they are kept as is to make it easier to backport changes.
```

## General style guidelines
Most of the style guidelines are enforced by clang format, but some additional comments are included here.

1. C style comments `/* comment */` can be used for both single and multi-line comments.
   C++ comments `//` can only be used for single line comments.
   Multi line comments should have the leading `*` align and the final `*/` should be on the same line as the last line of text.
   e.g.
```c
/* Blah Blah
 * Blah Blah. */
   ```
2. Comments should generally be used to describe behavior that is not obvious from reading the code itself.
   This includes complex behavior, why code was written the way it was and describing non-obvious behavior.
   Additionally, functions should be documented to explain all of the function's behavior without having to read the code.
1. Generally keep line lengths below 90 characters when reasonable, however there is no explicit line length enforcement.
   Use your best judgement for readability.
1. Use static functions when a function is only intended to be accessed from the same file.
   For historical reasons, some private functions are prefixed by `_`, and they are kept as is to make it easier to backport changes.

> 1. Use the boolean type for true/false values.
>    For historical reasons, some functions used the integer type, and they are kept as is to make it easier to backport changes.
> ## Naming conventions

Author: ```suggestion
   For historical reasons, some functions used the integer type, and they are kept as is to make it easier to backport changes.

## Naming conventions
```
